### ISSUE TITLE
AP exits silently on startup

### ISSUE THREAD

**Issue by chymian** [jbexta/AgentPilot#16]
I was trying to run the latest version on my debian 12, KDE  laptop.

first, there was no config file in the dir. of the AppImage, which crashed the app.
```
./AgentPilot-0.1.5.AppImage 
…
FileNotFoundError: [Errno 2] No such file or directory: '/home/guenter/tmp/AgentPilot.weg/configuration.yaml'
[3699489] Failed to execute script '__main__' due to unhandled exception!
```

since on linux, binarys and configurations have their "common places", I suggest following the [XDG Base Directory Specifications](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)

> $XDG_CONFIG_HOME defines the base directory relative to which user-specific configuration files should be stored. If $XDG_CONFIG_HOME is either not set or empty, a default equal to `$HOME/.config` should be used.

i.e. `$HOME/.config/AgentPilot/configuration.yaml`

> $XDG_DATA_HOME defines the base directory relative to which user-specific data files should be stored. If $XDG_DATA_HOME is either not set or empty, a default equal to `$HOME/.local/share` should be used.

that is for additional files of the APP, if neccessary. (i.e downloaded plugins)

for the local installed AppImages. they can go systemwide into `/usr/local/bin` or on userlevel into `$HOME/.local/bin`.
the latter is, where a AppImage like yours should be placed normally.

After having given the AppImage an own folder with cloned configuration file, the App starts for a short moment and exits silently.

I have attached a strace log, in case that helps.

[strace.log](https://github.com/jbexta/AgentPilot/files/13692445/strace.log)

**Comment by jbexta** [comment]
Hi, mine fails silently too, it really shouldnt and I'll add better exception handling so it's not silent.
I think it's an opengl dependency issue, if you run the appimage from terminal it would confirm it's the same problem (it says something about a missing file in the OS)

Right now the linux appimage is built on Ubuntu 18 which might be causing the problem, when I get chance I'll try build on an old Debian version and that might solve it.
For now the only way to get it running is to build from source on that machine following the instructions in the readme [How to build from source](https://github.com/jbexta/AgentPilot/blob/master/docs/guides/how_to_build.md)

But it's strange that there was no configuration file, I just downloaded the tar and it's there. I might get rid of this file as it's not really necessary anymore since there's a database

It doesn't have buld scripts yet so I'm manually building them, I might add platform specific builds when I get around to writing build scripts

**Comment by chymian** [comment]
@jbexta 
I did run it from a terminal and it was failing silently, hence the strace.

> But it's strange that there was no configuration file, I just downloaded the tar and it's there. I might get rid of this file as it's not really necessary anymore since there's a database

my bad, I copied only the binary and oversaw the config file.

on the other hand, it would be nice if you could support the mentioned XDG structure.
normaly binary and config/data files are stored at different places on linux.

**Comment by jbexta** [comment]
No worries, yea I'll get around to it eventually it's not finished yet but once it is or when I (or anyone) writes build scripts I'll do platform specific installers. The portable structure is easy to manage and deploy for now

**Comment by chymian** [comment]
I tried it again on a new account.

have found 2 culprits:
1. the silent exit I was faceing, has it's cause in the missing db-file. 
 pls. create all neccessary files on startup, if missing.

2. could get the APP up.
- played arround with custon API points (should point towards my LiteLLM-Proxy)
- it was possible to enable mem-gpt, but not avaiable
- after I tried these changes, the app crashes at startup:

```log
./AgentPilot-0.1.5.AppImage 
['/tmp/_MEIfbjrSI/base_library.zip', '/tmp/_MEIfbjrSI/lib-dynload', '/tmp/_MEIfbjrSI']
Traceback (most recent call last):
  File "agentpilot/__main__.py", line 271, in <module>
  File "agentpilot/__main__.py", line 267, in main
  File "agentpilot/gui.py", line 5256, in run
  File "agentpilot/gui.py", line 5061, in __init__
  File "agentpilot/gui.py", line 3548, in __init__
  File "agentpilot/context/base.py", line 67, in __init__
  File "agentpilot/context/base.py", line 75, in load
  File "agentpilot/context/base.py", line 156, in load_members
  File "agentpilot/agent/base.py", line 28, in __init__
  File "agentpilot/agent/base.py", line 117, in load_agent
Exception: Plugin "memgpt" not recognised
[1366346] Failed to execute script '__main__' due to unhandled exception!
Exception ignored in: <function Agent.__del__ at 0x7fa8737272e0>
Traceback (most recent call last):
  File "agentpilot/agent/base.py", line 56, in __del__
AttributeError: 'Agent' object has no attribute 'bg_task'
```

removing the db-file: app crashes silently again…
```bash
$ ./AgentPilot-0.1.5.AppImage ; echo $?
['/tmp/_MEI9z2Lr2/base_library.zip', '/tmp/_MEI9z2Lr2/lib-dynload', '/tmp/_MEI9z2Lr2']
0
```

hope this helps somehow.

#### Q: after I did set up my own custom provider/models I could not find them in the model-list to activate them?

**Comment by jbexta** [comment]
Do you have version 0.1.5 running on debian 12? What do you mean by new account?
Are you extracting the tar before opening the appimage? 
On the next release 0.1.6 any errors on startup won't be silent and memgpt will be available again with more settings.
I can't reproduce the model list bug, but I'm somewhere between 0.1.5 and 0.1.6. I thought I fixed that before 0.1.5. 
I'll build an alternative appimage for linux until the opengl problem is solved

**Comment by jbexta** [comment]
Is the custom api working ok for you apart from those issues? I've not tested those myself tbh

**Comment by chymian** [comment]
@jbexta 
- yes, 0.1.5 running on debian 12 KDE.
- new account, means fresh user account.
- extracting the tar: ROFL
- on start, I see the following errors:
```log
libGL error: MESA-LOADER: failed to open iris: /usr/lib/dri/iris_dri.so: Kann die Shared-Object-Datei nicht öffnen: Datei oder Verzeichnis nicht gefunden (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri, suffix _dri)
libGL error: failed to load driver: iris
```
the user is groupmember of `video` and `render`, even so, now dedicated GPU on my laptop and `libgl1-mesa-dri` is installed on my laptop.

**Comment by chymian** [comment]
> Is the custom api working ok for you apart from those issues? I've not tested that myself tbh

not really:
1. via litellm-Proxy:
  the proxy forwards GPT-Model-names to together.ai.
  LiteLLM: Current Version = 1.15.1
 
2. using together AI from within AG.

both create the same error:
the answer to a query is shown in the window, till the little error window pops up. aknowledgeing the error, the answer disapears.

![2023-12-30_14-37AG_Answer-error](https://github.com/jbexta/AgentPilot/assets/1899961/517e1ba7-89eb-4ef4-acff-3522615691f5)

![2023-12-30_14-38AG_Answer missing](https://github.com/jbexta/AgentPilot/assets/1899961/3fe9e481-926c-43fe-82a4-c0c1aede50a1)


## incomplete model list

since the modellist of every provider is updated frequently, the build-in model list is somehow "always" out of date.
and the list is quite confused, no alpapetical sort order.

May I suggest, to first choose the provider, then query and show the latest modellist.

**Comment by jbexta** [comment]
Ok I thought I fixed the last error you mentioned on 0.1.5 but it might have been just after. I'm gona push 0.1.6 this week and that error should be gone. Also I will build on Debian instead of ubuntu to see if that solves the opengl issue.
I've added API headers to the model list so it's easier to find models, it's ordered first by API name, then by the model alias, so it's alphabetical but within each API. Also I'll probably add the ability to favourite APIs/models so the model list only shows the favourited ones
